lib/cocoapods/installer/analyzer.rb

```diff
@@ -231,6 +232,29 @@ def validate_lockfile_version!
         end
       end
 
+      # Downloads all the podspecs and versions for all the pods in the lockfile
+      # for CDN repositories.
+      #
+      # @note   This is an optimization step to reduce delays due to making
+      #         multiple CDN requests serially. We make all the requests
+      #         in parallel.
+      def pre_cache_cdn_source_from_lockfile
+        return if lockfile.nil?
+
+        lockfile.pods_by_spec_repo.map do |source_url, pods|
+          source = config.sources_manager.find_or_create_source_with_url(source_url)
+          next if pods.nil?
+          next unless source.is_a?(Pod::CDNSource)
+
+          threads = pods.map do |pod|
+            Thread.new do
```

Comment:
Instead of spinning up a thread for each pod, it might make sense to use a fixed-sized thread pool, enqueue pods for each source on it, and then wait for it to drain at the end of the method?

Reply:
Here's the thing: this parallelization is here to take advantage of the fact that most of the delay in podspec retrieval is due to CDN latency. The threads will be waiting for the request to complete for most of their lifetime. HTTP pipelining itself is taken care of by the HTTP client.

In addition to this parallelization, there is another happening in CDNSource that retrieves all the podspecs once a version list is requested.

I believe there's no inherent problem in spawning all these threads, but occasionally, jsDelivr will error out with an "connection refused". This doesn't happen very often, and never on the happy, cached install.